src/settings/factory_reset.c

```diff
@@ -0,0 +1,294 @@
+      memcpy(&wallet_list_out->wallet[wallet_list_out->count],
+             &wallet_list1->wallet[index],
+             sizeof(wallet_metadata_t));
+      wallet_list_out->count += 1;
+    }
+  }
+
+  return;
+}
+
+static bool filter_uncommon_wallet(bool match_found) {
+  return (!match_found);
+}
+
+static bool filter_common_wallet(bool match_found) {
+  return match_found;
+}
+
+static bool get_wallet_list_from_two_cards(wallet_list_t *wallet_list) {
+  wallet_list_t wallets_in_card1 = {0};
+  wallet_list_t wallets_in_card2 = {0};
+
+  card_fetch_wallet_list_config_t configuration = {0};
+  card_fetch_wallet_list_response_t response = {0};
+
+  configuration.acceptable_cards = 0xF;
+  configuration.heading = ui_text_tap_1_2_cards;
+  configuration.msg = ui_text_place_card_below;
+  configuration.skip_card_removal = false;
+
+  response.wallet_list = &wallets_in_card1;
+  response.tapped_card = 0;
+
+  if (!card_fetch_wallet_list(&configuration, &response)) {
+    return false;
+  }
+
+  // Do not accept the same card again
+  configuration.acceptable_cards ^= response.tapped_card;
+  configuration.heading = ui_text_tap_2_2_cards;
+  configuration.skip_card_removal = true;
+
+  response.wallet_list = &wallets_in_card2;
+  response.tapped_card = 0;
+
+  if (!card_fetch_wallet_list(&configuration, &response)) {
+    return false;
+  }
+
+  delay_scr_init(ui_text_processing, DELAY_TIME);
+
+  filter_wallet_list(
+      &wallets_in_card1, &wallets_in_card2, wallet_list, filter_common_wallet);
+  return true;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void factory_reset(void) {
+  wallet_list_t wallets_in_cards = {0};
+  wallet_list_t wallets_in_vault = {0};
+
+  uint8_t valid_wallets = get_valid_wallet_meta_data_list(&wallets_in_vault);
+
+  if (0 < valid_wallets &&
+      !core_scroll_page(NULL, ui_text_factory_reset_instruction, NULL)) {
+    return;
+  }
+
+  if (!core_scroll_page(NULL, ui_text_confirm_factory_reset, NULL)) {
+    return;
+  }
+
+  // If there is atleast one valid wallet on the device, ensure that it is
+  // present in atleast 2 X1 cards
+  if (0 < valid_wallets) {
+    if (!get_wallet_list_from_two_cards(&wallets_in_cards)) {
+      return;
+    }
+
+    wallet_list_t uncommon_wallets = {0};
+    filter_wallet_list(&wallets_in_vault,
+                       &wallets_in_cards,
+                       &uncommon_wallets,
+                       filter_uncommon_wallet);
+
+    if (0 < uncommon_wallets.count) {
+      // Display first wallet not present in cards - just to keep it
+      // consistent with current UX
+      char msg[64];
+      const char *msg_list[3] = {
+          msg, ui_text_reset_exit[0], ui_text_reset_exit[1]};
+
+      snprintf(msg,
+               sizeof(msg),
+               UI_TEXT_FACTORY_RESET_ERROR,
+               uncommon_wallets.wallet[0].name);
```

Comment:
Consider updating the message to indicate that multiple wallets are not present on either 1 or more than 1 card.
Let's change this in the draw.io designs first and take this change up as a separate improvement.

Reply:
Should add a todo just to make sure we don't miss this for too long.

Reply:
Added